src/machine/machine-varlink.c

```diff
@@ -375,3 +378,188 @@ int vl_method_kill(sd_varlink *link, sd_json_variant *parameters, sd_varlink_met
+                case MACHINE_OPEN_MODE_LOGIN:
+                        return (const char**) STRV_MAKE("machine", machine_name, "verb", "login");
+                case MACHINE_OPEN_MODE_SHELL:
+                        assert(user);
+                        assert(path);
+                        assert(command_line);
+
+                        return (const char**) STRV_MAKE(
+                                        "machine", machine_name,
+                                        "verb", "shell",
+                                        "user", user,
+                                        "program", path,
+                                        "command_line", command_line);
+                default:
+                        assert_not_reached();
+        }
+}
+
+int vl_method_open(sd_varlink *link, sd_json_variant *parameters, sd_varlink_method_flags_t flags, void *userdata) {
+        static const sd_json_dispatch_field dispatch_table[] = {
+                VARLINK_DISPATCH_MACHINE_LOOKUP_FIELDS(MachineOpenParameters),
+                { "mode", SD_JSON_VARIANT_STRING, json_dispatch_machine_open_mode, offsetof(MachineOpenParameters, mode), 0 },
+                { "user", SD_JSON_VARIANT_STRING, sd_json_dispatch_const_string,   offsetof(MachineOpenParameters, user), 0 },
+                { "path", SD_JSON_VARIANT_STRING, sd_json_dispatch_const_string,   offsetof(MachineOpenParameters, path), 0 },
+                { "args", SD_JSON_VARIANT_ARRAY,  sd_json_dispatch_strv,           offsetof(MachineOpenParameters, args), 0 },
+                { "env",  SD_JSON_VARIANT_ARRAY,  sd_json_dispatch_strv,           offsetof(MachineOpenParameters, env),  0 },
+                VARLINK_DISPATCH_POLKIT_FIELD,
+                {}
+        };
+
+        Manager *manager = ASSERT_PTR(userdata);
+        _cleanup_close_ int master = -EBADF;
+        _cleanup_(machine_open_paramaters_done) MachineOpenParameters p = { .pidref = PIDREF_NULL };
+        _cleanup_(sd_json_variant_unrefp) sd_json_variant *v = NULL;
+        _cleanup_free_ char *pty_name = NULL, *command_line = NULL;
+        _cleanup_strv_free_ char **args = NULL;
+        const char *user = NULL, *path = NULL; /* gcc complains about uninitialized variables */
+        Machine *machine;
+        int r;
+
+        assert(link);
+        assert(parameters);
+
+        r = sd_varlink_dispatch(link, parameters, dispatch_table, &p);
+        if (r != 0)
+                return r;
+
+        if (p.mode == MACHINE_OPEN_MODE_SHELL) {
+                if (!isempty(p.user) && !valid_user_group_name(p.user, VALID_USER_RELAX))
+                        return sd_varlink_error_invalid_parameter_name(link, "user");
+                if (!isempty(p.path) && !path_is_absolute(p.path))
+                        return sd_varlink_error_invalid_parameter_name(link, "path");
+                // TODO(ikruglov): args validation?
+                if (!strv_isempty(p.env) && !strv_env_is_valid(p.env))
```

Comment:
i sense that validating env blocks is so generic, that we should have json_dispatch_strv_env() for this too. we'll need it once we add varlink support for service activation in pid 1

Reply:
heck, we already have such a function in src/shared/user-record.c: json_dispatch_environment()

can you move that to src/libsystemd/sd-json/json-util.[ch] and maybe rename it json_dispatch_strv_env() since we apparently otherwise call these strv with env assignments strv_env